core/state_transition.go

```diff
@@ -366,9 +374,77 @@ func (st *StateTransition) applyEditValidatorTx(ev *staking.EditValidator, block
 }
 
 func (st *StateTransition) applyDelegateTx(delegate *staking.Delegate) error {
+	if !st.state.IsValidator(delegate.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+	wrapper := st.state.GetStakingInfo(delegate.ValidatorAddress)
+	if wrapper == nil {
+		return errValidatorNotExist
+	}
+
+	stateDB := st.state
+	delegatorExist := false
+	for i := range wrapper.Delegations {
+		delegation := wrapper.Delegations[i]
+		if bytes.Equal(delegation.DelegatorAddress.Bytes(), delegate.DelegatorAddress.Bytes()) {
+			delegatorExist = true
+			if CanTransfer(stateDB, delegate.DelegatorAddress, delegate.Amount) {
+				delegation.Amount.Add(delegation.Amount, delegate.Amount)
+				err := stateDB.UpdateStakingInfo(wrapper.Validator.Address, wrapper)
+				if err == nil {
+					// TODO: use undelegated token in locking period
+					stateDB.SubBalance(delegate.DelegatorAddress, delegate.Amount)
```

Comment:
is it ever possible to have a negative balance?

Reply:
Good question. It's possible. And the protocol should eliminate those transactions in the very beginning.